### PR TITLE
Add missing capabilities for the GraphQL endpoint

### PR DESCRIPTION
**Pull request by msprunck**
Related to threatgrid/iroh#2006

A `wrap-authenticated` middleware has been added to check that the current user is authenticated
even if no capabilities are specified for a protected route.

### CODE REVIEW

File: src/ctia/graphql/routes.clj [hunk 1]
```diff
@@ -31,6 +31,40 @@
...
         :summary "EXPERIMENTAL: Executes a Relay compatible GraphQL query"
+        :capabilities
+        #{:list-campaigns
+          :read-actor
+          :read-malware
+          :read-exploit-target
+          :read-attack-pattern
+          :read-judgement
+          :read-sighting
+                  :list-sightings
```
Comment (yogsototh) [inline src/ctia/graphql/routes.clj]: bad indentation.

File: src/ctia/http/handler.clj [hunk 2]
```diff
@@ -44,7 +44,7 @@
   When an Observable with a malicious Verdict is seen, it can be recorded as
   a Sighting, and the Relations that Observable had with other Observables can
   be recorded as well.
-
+  d
```
Comment (yogsototh) [inline src/ctia/http/handler.clj]: certainly a mistake.

File: src/ctia/graphql/routes.clj (hunk 1 above)
Comment (gbuisson) [inline src/ctia/graphql/routes.clj]: why a TAB here ?

File: src/ctia/http/handler.clj [hunk 3]
```diff
@@ -131,20 +131,23 @@
...
-                     incident-casebook-link-route)
-                    bundle-routes
-                    observable-routes
-                    metrics-routes
-                    properties-routes
-                    graphql-routes
-                    version-routes))
+                    ;; The order is important here for version-routes
+                    ;; must be before the middleware fn
+                    version-routes
```
Comment (yogsototh) [inline src/ctia/http/handler.clj]: ah....

File: src/ctia/http/handler.clj (hunk 2 above)
Comment (gbuisson) [inline src/ctia/http/handler.clj]: typo

File: test/ctia/http/routes/graphql_test.clj [hunk 4]
```diff
@@ -650,3 +660,6 @@
              (is (empty? errors) "No errors")
              (is (= "indicator" (get-in data [:sighting :relationships :nodes 0
                                               :target_entity :type]))))))))))
+
+(deftest graphql-access-without-capabilities
```
Comment (gbuisson) [inline test/ctia/http/routes/graphql_test.clj]: and ?

File: src/ctia/http/middleware/auth.clj [hunk 5]
```diff
@@ -32,6 +32,14 @@
 (defn wrap-authentication [handler]
   (testable-wrap-authentication handler @auth-service))
 
+(defn wrap-authenticated [handler]
+  (fn [request]
+    (let [auth-identity (:identity request)]
+      (if (and (not (nil? auth-identity))
+               (auth/authenticated? auth-identity))
+        (handler request)
+        (http-response/forbidden! {:message "Only authenticated users allowed"})))))
```
Comment (yogsototh) [inline src/ctia/http/middleware/auth.clj]: The fact that this is not configurable make it impossible to spawn a fully public instance.
I think it's ok for now. But I believe someone might want to provide a fully open to the public instance even for non identified users.

File: test/ctia/http/routes/graphql_test.clj [hunk 6]
```diff
@@ -650,3 +660,6 @@
...
+  )
```
Comment (yogsototh) [inline test/ctia/http/routes/graphql_test.clj]: unfinished here. Certainly a mistake.

File: src/ctia/http/middleware/auth.clj (hunk 5 above)
Comment (msprunck) [reply src/ctia/http/middleware/auth.clj]: It's possible with the `allow-all` authentication mode. It always returns `true` to  `authenticated?`

File: src/ctia/graphql/routes.clj (hunk 1 above)
Comment (msprunck) [reply src/ctia/graphql/routes.clj]: I prefer tabs over spaces!

**Comment by hshahman** [comment]
Verified in rel-v1.4

**Pull request merged**